// ngx-filepond/esm2020/lib/ngx-filepond.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FilePondComponent } from './ngx-filepond.component';
import * as i0 from "@angular/core";
export { registerPlugin } from 'filepond';
export class FilePondModule {
}
FilePondModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.1", ngImport: i0, type: FilePondModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FilePondModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.1", ngImport: i0, type: FilePondModule, declarations: [FilePondComponent], imports: [CommonModule], exports: [FilePondComponent] });
FilePondModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.1", ngImport: i0, type: FilePondModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.1", ngImport: i0, type: FilePondModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [FilePondComponent],
                    exports: [FilePondComponent],
                }]
        }] });
