// @cloudinary/ng/esm2015/lib/cloudinary.module.js
import { NgModule } from '@angular/core';
import { CloudinaryImageComponent } from './cloudinary-image.component';
import { CloudinaryVideoComponent } from './cloudinary-video.component';
import * as i0 from "@angular/core";
export class CloudinaryModule {
}
CloudinaryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: CloudinaryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
CloudinaryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: CloudinaryModule, declarations: [CloudinaryImageComponent, CloudinaryVideoComponent], exports: [CloudinaryImageComponent, CloudinaryVideoComponent] });
CloudinaryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: CloudinaryModule, imports: [[]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.17", ngImport: i0, type: CloudinaryModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [],
                    declarations: [CloudinaryImageComponent, CloudinaryVideoComponent],
                    exports: [CloudinaryImageComponent, CloudinaryVideoComponent]
                }]
        }] });
